src/Lib.P2P/SecureCommunication/Secio1Stream.cs

```diff
@@ -0,0 +1,270 @@
+        }
+
+        /// <inheritdoc />
+        public override long Seek(long offset, SeekOrigin origin) { throw new NotSupportedException(); }
+
+        /// <inheritdoc />
+        public override void SetLength(long value) { throw new NotSupportedException(); }
+
+        /// <inheritdoc />
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+#pragma warning disable VSTHRD002 
+            return ReadAsync(buffer, offset, count).GetAwaiter().GetResult();
+#pragma warning restore VSTHRD002 
+        }
+
+        /// <inheritdoc />
+        public override async Task<int> ReadAsync(byte[] buffer,
+            int offset,
+            int count,
+            CancellationToken cancellationToken)
+        {
+            var total = 0;
+            while (count > 0)
+
+                // Does the current packet have some unread data?
+                if (inBlock != null && inBlockOffset < inBlock.Length)
+                {
+                    var n = Math.Min(inBlock.Length - inBlockOffset, count);
+                    Array.Copy(inBlock, inBlockOffset, buffer, offset, n);
+                    total += n;
+                    count -= n;
+                    offset += n;
+                    inBlockOffset += n;
+                }
+
+                // Otherwise, wait for a new block of data.
+                else
+                {
+                    inBlock = await ReadPacketAsync(cancellationToken);
+                    inBlockOffset = 0;
+                }
+
+            return total;
+        }
+
+        /// <summary>
+        ///   Read an encrypted and signed packet.
+        /// </summary>
+        /// <returns>
+        ///   The plain text as an array of bytes.
+        /// </returns>
+        /// <remarks>
+        ///   A packet consists of a [uint32 length of packet | encrypted body | hmac signature of encrypted body].
+        /// </remarks>
+        private async Task<byte[]> ReadPacketAsync(CancellationToken cancel)
+        {
+            var lengthBuffer = await ReadPacketBytesAsync(4, cancel).ConfigureAwait(false);
+            var length =
+                ((int) lengthBuffer[0] << 24) |
+                ((int) lengthBuffer[1] << 16) |
+                ((int) lengthBuffer[2] << 8) |
+                (int) lengthBuffer[3];
+            if (length <= outHmac.GetMacSize())
+                throw new InvalidDataException($"Invalid secio packet length of {length}.");
+
+            var encryptedData = await ReadPacketBytesAsync(length - outHmac.GetMacSize(), cancel).ConfigureAwait(false);
+            var signature = await ReadPacketBytesAsync(outHmac.GetMacSize(), cancel).ConfigureAwait(false);
+
+            var hmac = outHmac;
+            var mac = new byte[hmac.GetMacSize()];
+            hmac.Reset();
+            hmac.BlockUpdate(encryptedData, 0, encryptedData.Length);
+            hmac.DoFinal(mac, 0);
+            if (!signature.SequenceEqual(mac))
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)